gen/plugin.go

```diff
@@ -0,0 +1,348 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/thriftrw/thriftrw-go/compile"
+	"github.com/thriftrw/thriftrw-go/plugin/api"
+)
+
+type serviceName string
+
+type generateServiceBuilder struct {
+	api.GenerateServiceRequest
+
+	importer thriftPackageImporter
+
+	nextModuleID  api.ModuleID
+	nextServiceID api.ServiceID
+
+	// ThriftFile -> Module ID
+	moduleIDs map[string]api.ModuleID
+
+	// ThriftFile -> Service name -> Service ID
+	serviceIDs map[string]map[serviceName]api.ServiceID
+
+	// To ensure there are no duplicates
+	rootServices map[api.ServiceID]struct{}
+}
+
+func newGenerateServiceBuilder(i thriftPackageImporter) *generateServiceBuilder {
+	return &generateServiceBuilder{
+		GenerateServiceRequest: api.GenerateServiceRequest{
+			RootServices: make([]api.ServiceID, 0, 10),
+			Services:     make(map[api.ServiceID]*api.Service),
+			Modules:      make(map[api.ModuleID]*api.Module),
+		},
+		importer:      i,
+		nextModuleID:  1,
+		nextServiceID: 1,
+		moduleIDs:     make(map[string]api.ModuleID),
+		serviceIDs:    make(map[string]map[serviceName]api.ServiceID),
+		rootServices:  make(map[api.ServiceID]struct{}),
+	}
+}
+
+func (g *generateServiceBuilder) Build() *api.GenerateServiceRequest {
+	return &g.GenerateServiceRequest
+}
+
+// AddRootService adds a service as a root service to this request.
+func (g *generateServiceBuilder) AddRootService(spec *compile.ServiceSpec) (api.ServiceID, error) {
+	id, err := g.addService(spec)
+	if err != nil {
+		return id, err
+	}
+
+	if _, alreadyAdded := g.rootServices[id]; !alreadyAdded {
```

Comment:
No. `RootServices` is just a list of pointers to services for which code should be generated. You won't generate code for the same service twice because it would end up overwriting itself.